### PR TITLE
Fix table rate failing for zip+4 address #17770

### PR DESCRIPTION
**Pull request by magently**
### Description
Currently when a customer uses an American ZIP+4 postcode, the table rate shipping method fails to return a quote.

### Fixed Issues (if relevant)
1. magento/magento2#17770

### Manual testing scenarios
1. Create a Product with weight >= 1.
2. Enable the Table shipping rate 
3. Import the rate CSV file:
`Country,Region/State,Zip/Postal Code,Weight (and above),Shipping Price`
`USA,*,12345,1,2`
4. Create a customer on the frontend with a zip code in the format ZIP+4 (ex: 12345-1234).
5. Add Product to your cart
6. Go to checkout

### Expected result

1. The table rate shipping method should ignore the extension and  appear with a 2$ shipping fee.

### Actual result

1. The table rate shipping is not present.

The issue is now fixed. If the customer uses a ZIP+4 code (ex: 12345-1234) in case a rate for an exact match (12345-1234) is not found in the table there will be a fallback check for 12345.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=18166) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @magently. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: app/code/Magento/OfflineShipping/Model/ResourceModel/Carrier/Tablerate/RateQuery.php [hunk 1]
```diff
@@ -51,7 +52,7 @@ public function prepareSelect(\Magento\Framework\DB\Select $select)
                 "dest_country_id = '0' AND dest_region_id = 0 AND dest_zip = '*'",
                 "dest_country_id = :country_id AND dest_region_id = 0 AND dest_zip = ''",
                 "dest_country_id = :country_id AND dest_region_id = 0 AND dest_zip = :postcode",
-                "dest_country_id = :country_id AND dest_region_id = 0 AND dest_zip = '*'"
```
Comment (slavvka) [inline app/code/Magento/OfflineShipping/Model/ResourceModel/Carrier/Tablerate/RateQuery.php:54]: I am wondering why did you remove the line with the star?

File: app/code/Magento/OfflineShipping/Model/ResourceModel/Carrier/Tablerate/RateQuery.php [hunk 2]
```diff
@@ -112,4 +114,18 @@ public function getRequest()
     {
         return $this->request;
     }
+
+    /**
+     * Returns the entire postcode if it contains no dash
+     * or the part of it prior to the dash in the other case
+     * @return string
+     */
+    public function getDestPostcodePrefix()
```
Comment (slavvka) [inline app/code/Magento/OfflineShipping/Model/ResourceModel/Carrier/Tablerate/RateQuery.php]: I see no reason to make this method public. Do you?

**Review by slavvka — CHANGES_REQUESTED** [review]

File: app/code/Magento/OfflineShipping/Model/ResourceModel/Carrier/Tablerate/RateQuery.php (hunk 1 above)
Comment (magently) [reply app/code/Magento/OfflineShipping/Model/ResourceModel/Carrier/Tablerate/RateQuery.php:54]: If you look at lines 49 and 54 in original code, you will see that there were two identical conditions in this OR statement. So we removed one.

Comment (slavvka) [reply app/code/Magento/OfflineShipping/Model/ResourceModel/Carrier/Tablerate/RateQuery.php:54]: @magently yes, I missed that. You are right

**Review by slavvka — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @slavvka, thank you for the review.
[ENGCOM-3057](https://jira.corp.magento.com/browse/ENGCOM-3057) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
Hi @magently. Thank you for your contribution. 
We will aim to release these changes as part of 2.2.8. 
Please check the release notes for final confirmation.

Please, consider to port this solution to 2.3 release line.
You may use [Porting tool](https://github.com/magento/magento2/wiki/Simple-way-to-port-your-solutions-automatically) to port commits automatically.
